File: sdks/go/corthan/logger.go
package corthan

import (
	"log"
	"os"
)

// Logger defines an interface for structured logging inside the SDK.
type Logger interface {
	Debug(msg string, keysAndValues ...interface{})
	Info(msg string, keysAndValues ...interface{})
	Warn(msg string, keysAndValues ...interface{})
	Error(msg string, keysAndValues ...interface{})
}

// NoOpLogger is a logger that discards all log messages.
type NoOpLogger struct{}

func (n *NoOpLogger) Debug(msg string, keysAndValues ...interface{}) {}
func (n *NoOpLogger) Info(msg string, keysAndValues ...interface{})  {}
func (n *NoOpLogger) Warn(msg string, keysAndValues ...interface{})  {}
func (n *NoOpLogger) Error(msg string, keysAndValues ...interface{}) {}

// StdLogger is a basic logger that writes to stdout/stderr.
type StdLogger struct {
	debugLog *log.Logger
	infoLog  *log.Logger
	warnLog  *log.Logger
	errLog   *log.Logger
	verbose  bool
}

// NewStdLogger instantiates a standard stdout/stderr logger.
func NewStdLogger(verbose bool) *StdLogger {
	return &StdLogger{
		debugLog: log.New(os.Stdout, "[corthan-sdk] DEBUG: ", log.LstdFlags),
		infoLog:  log.New(os.Stdout, "[corthan-sdk] INFO: ", log.LstdFlags),
		warnLog:  log.New(os.Stdout, "[corthan-sdk] WARN: ", log.LstdFlags),
		errLog:   log.New(os.Stderr, "[corthan-sdk] ERROR: ", log.LstdFlags),
		verbose:  verbose,
	}
}

func (s *StdLogger) Debug(msg string, keysAndValues ...interface{}) {
	if s.verbose {
		s.debugLog.Println(msg, keysAndValues)
	}
}

func (s *StdLogger) Info(msg string, keysAndValues ...interface{}) {
	s.infoLog.Println(msg, keysAndValues)
}

func (s *StdLogger) Warn(msg string, keysAndValues ...interface{}) {
	s.warnLog.Println(msg, keysAndValues)
}

func (s *StdLogger) Error(msg string, keysAndValues ...interface{}) {
	s.errLog.Println(msg, keysAndValues)
}

File: sdks/node/src/logger.ts
/**
 * Structured logger interface used by the SDK.
 */
export interface Logger {
  /** Log a message at the debug level. */
  debug(msg: string, ...args: any[]): void;
  /** Log a message at the info level. */
  info(msg: string, ...args: any[]): void;
  /** Log a message at the warn level. */
  warn(msg: string, ...args: any[]): void;
  /** Log a message at the error level. */
  error(msg: string, ...args: any[]): void;
}

/**
 * NoOpLogger discards all logging output.
 */
export class NoOpLogger implements Logger {
  debug(): void {}
  info(): void {}
  warn(): void {}
  error(): void {}
}

/**
 * StdLogger outputs logs to the standard stdout/stderr streams.
 */
export class StdLogger implements Logger {
  /**
   * @param verbose If true, debug logs are written to stdout.
   */
  constructor(private verbose: boolean = false) {}

  debug(msg: string, ...args: any[]): void {
    if (this.verbose) {
      console.log(`[corthan-sdk] DEBUG: ${msg}`, ...args);
    }
  }

  info(msg: string, ...args: any[]): void {
    console.log(`[corthan-sdk] INFO: ${msg}`, ...args);
  }

  warn(msg: string, ...args: any[]): void {
    console.warn(`[corthan-sdk] WARN: ${msg}`, ...args);
  }

  error(msg: string, ...args: any[]): void {
    console.error(`[corthan-sdk] ERROR: ${msg}`, ...args);
  }
}
